src/libse/AudioToText/AudioToTextPostProcessor.cs

```diff
@@ -226,46 +232,48 @@ public Subtitle AddPeriods(Subtitle inputSubtitle, string language)
+            };
+
+            for (var index = 0; index < subtitle.Paragraphs.Count; index++)
             {
                 var paragraph = subtitle.Paragraphs[index];
-                var next = subtitle.Paragraphs[index + 1];
-                if (next.StartTime.TotalMilliseconds - paragraph.EndTime.TotalMilliseconds > SetPeriodIfDistanceToNextIsMoreThan &&
-                    !paragraph.Text.EndsWith('.') &&
-                    !paragraph.Text.EndsWith('!') &&
-                    !paragraph.Text.EndsWith('?') &&
-                    !paragraph.Text.EndsWith(',') &&
-                    !paragraph.Text.EndsWith(':'))
+                var next = subtitle.GetParagraphOrDefault(index + 1) ?? defaultParagraph;
+
+                // paragraph is already closed
+                if (paragraph.Text.EndsWith('.') ||
+                    paragraph.Text.EndsWith('!') ||
+                    paragraph.Text.EndsWith('?') ||
+                    paragraph.Text.EndsWith(',') ||
+                    paragraph.Text.EndsWith(':'))
+                {
+                    continue;
+                }
+
+                var timeDistanceToNextLine = next.StartTime.TotalMilliseconds - paragraph.EndTime.TotalMilliseconds;
+                if (timeDistanceToNextLine > SetPeriodIfDistanceToNextIsMoreThan)
+                {
+                    paragraph.Text += ".";
+                }
+                else
                 {
-                    if (next.StartTime.TotalMilliseconds - paragraph.EndTime.TotalMilliseconds > SetPeriodIfDistanceToNextIsMoreThanAlways)
+                    if (TwoLetterLanguageCode == "en" &&
+                        (englishSkipLastWords.Contains(GetLastWord(paragraph.Text)) ||
+                         englishSkipFirstWords.Contains(GetFirstWord(next.Text))))
                     {
-                        paragraph.Text += ".";
+                        continue;
                     }
-                    else
-                    {
-                        var lastWord = GetLastWord(paragraph.Text);
-                        var nextFirstWord = GetFirstWord(next.Text);
-                        if (TwoLetterLanguageCode == "en" && (englishSkipLastWords.Contains(lastWord) || englishSkipFirstWords.Contains(nextFirstWord)))
-                        {
-                            continue;
-                        }
 
-                        paragraph.Text += ".";
-                    }
+                    paragraph.Text += ".";
                 }
             }
 
-            var last = subtitle.GetParagraphOrDefault(subtitle.Paragraphs.Count - 1);
-            if (last != null &&
-                !last.Text.EndsWith('.') &&
-                !last.Text.EndsWith('!') &&
-                !last.Text.EndsWith('?') &&
-                !last.Text.EndsWith(',') &&
-                !last.Text.EndsWith(':'))
-            {
-                subtitle.Paragraphs[subtitle.Paragraphs.Count - 1].Text += ".";
-            }
```

Comment:
This code is unnecessary as we can use "default paragraph" (see line 238) with start-time of the last paragraph + SetPeriodIfDistanceToNextIsMoreThanAlways+1 and let the for-loop goes to `subtitle.Paragraph.Count` instead of  `subtitle.Paragraph.Count-1`.

When we `subtitle.GetParagraphOrDefault(index + 1)` if null is returned then we can be sure that the current paragraph is the last and just return the default  `defaultParagraph`